### PR TITLE
feat: blocked user conversation [WPB-6259]

### PR DESCRIPTION
**Pull request by PatrykBuniX**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6259" title="WPB-6259" target="_blank">StoryWPB-6259</a>    [Web] 1:1 Conversations with blocked users should be accessible still
  </td></table>
  <br />
 

## Description

This PR adds the ability to keep the conversation with a blocked user in a list and also changes how blocked user is indicated in the app.

## Screenshots/Screencast (for UI changes)

1:1 conversation view:
Screenshot 2024-04-12 at 11 24 55

Search:
Screenshot 2024-04-12 at 11 25 13

User details in the right sidebar:
Screenshot 2024-04-12 at 11 25 26

Indication in a group conversation with a blocked user:
Screenshot 2024-04-12 at 11 25 41

## Checklist

- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [ ] If it is a core feature, unit tests have been added;

### CODE REVIEW

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import {CONVERSATION_READONLY_STATE} from 'src/script/conversation/ConversationRepository';
+import {Conversation} from 'src/script/entity/Conversation';
+import {useKoSubscribableChildren} from 'Util/ComponentUtil';
+import {t} from 'Util/LocalizerUtil';
+
+interface ReadOnlyConversationMessageProps {
+  reloadApp: () => void;
+  conversation: Conversation;
+}
+export const ReadOnlyConversationMessage: FC<ReadOnlyConversationMessageProps> = ({conversation, reloadApp}) => {
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx:40]:
```suggestion
export const ReadOnlyConversationMessage = ({conversation, reloadApp}: ReadOnlyConversationMessageProps) => {
```

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  if (user.isBlocked()) {
+    return (
+      <ReadOnlyConversationMessageBase>
+        <span
+          dangerouslySetInnerHTML={{
+            __html: t('conversationWithBlockedUser'),
+          }}
+        />
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Hmm.. Are You sure to use there dangerouslySetInnerHTML ?

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  if (readOnlyState) {
+    switch (readOnlyState) {
+      case CONVERSATION_READONLY_STATE.READONLY_ONE_TO_ONE_OTHER_UNSUPPORTED_MLS:
+        return (
+          <ReadOnlyConversationMessageBase>
+            <span
+              dangerouslySetInnerHTML={{
+                __html: t('otherUserNotSupportMLSMsg', user.name()),
+              }}
+            />
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Are You sure to use there dangerouslySetInnerHTML ?

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+          </ReadOnlyConversationMessageBase>
+        );
+      case CONVERSATION_READONLY_STATE.READONLY_ONE_TO_ONE_SELF_UNSUPPORTED_MLS:
+        return (
+          <ReadOnlyConversationMessageBase>
+            <span
+              dangerouslySetInnerHTML={{
+                __html: t('selfNotSupportMLSMsgPart1', user.name()),
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Are You sure to use there dangerouslySetInnerHTML ?

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                variant={LinkVariant.PRIMARY}
+                data-uie-name="do-update-mls"
+              >
+                {t('downloadLatestMLS')}
+              </Link>{' '}
+              <span
+                dangerouslySetInnerHTML={{
+                  __html: t('selfNotSupportMLSMsgPart2'),
+                }}
+              />
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Are You sure to use there dangerouslySetInnerHTML ?

File: src/script/components/ParticipantItemContent/ParticipantItemContent.tsx [hunk 6]
```diff
@@ -81,6 +82,11 @@ export const ParticipantItemContent = ({
               showAvailability={showAvailabilityState && selfInTeam}
             >
               <UserVerificationBadges user={participant} groupId={groupId} />
+              {participant.isBlocked() && (
```
Comment (przemvs) [inline src/script/components/ParticipantItemContent/ParticipantItemContent.tsx]: I think You should use isBlocked from useKoSubscribableChildren .

File: src/script/view_model/ListViewModel.ts [hunk 7]
```diff
@@ -443,9 +449,20 @@ export class ListViewModel {
...
-    await this.actionsViewModel.blockUser(userEntity, hideConversation, nextConversationEntity);
+
+    if (!userEntity) {
+      return;
+    }
+
+    await this.actionsViewModel.blockUser(userEntity);
+  };
+
+  clickToUnblock = async (conversationEntity: Conversation): Promise<void> => {
```
Comment (przemvs) [inline src/script/view_model/ListViewModel.ts:460]: Maybe rename it to unblockUser ? :)

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx (hunk 2 above)
Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Right, this one doesn't use any html elements.

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx (hunk 5 above)
Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Right, this one doesn't use any html elements.

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx (hunk 3 above)
Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: This is the only way we can currently inject user name into the template.

File: src/script/components/ParticipantItemContent/ParticipantItemContent.tsx (hunk 6 above)
Comment (PatrykBuniX) [reply src/script/components/ParticipantItemContent/ParticipantItemContent.tsx]: Not really, as at that level the participant is either Service or User, isBlocked doesn't exist on Service entity.

File: src/script/view_model/ListViewModel.ts (hunk 7 above)
Comment (PatrykBuniX) [reply src/script/view_model/ListViewModel.ts:460]: There was `clickToBlock` and other actions named like this, I think it's better to stick to the existing convention.

File: src/script/components/ParticipantItemContent/ParticipantItemContent.tsx (hunk 6 above)
Comment (atomrc) [reply src/script/components/ParticipantItemContent/ParticipantItemContent.tsx]: But that's valid point though, in this case the UI will not rerender when the `isBlocked` value changes. 

We would need to find a way to `subscribe` using `useKoSubscribableChildren` at the top level. 

Maybe something like 

```
const {isBlocked} = useKoSubscribableChildren(!isService(user) ? user : {isBlocked: ko.observable(false)})

```

File: src/script/connection/ConnectionRepository.ts [hunk 8]
```diff
@@ -158,15 +158,8 @@ export class ConnectionRepository {
...
    */
-  public async blockUser(
-    userEntity: User,
-    hideConversation: boolean = false,
-    nextConversationEntity?: Conversation,
-  ): Promise<void> {
+  public async blockUser(userEntity: User): Promise<void> {
     await this.updateStatus(userEntity, ConnectionStatus.BLOCKED);
-    if (hideConversation) {
-      amplify.publish(WebAppEvents.CONVERSATION.SHOW, nextConversationEntity, {});
```
Comment (atomrc) [inline src/script/connection/ConnectionRepository.ts:168]: has this moved somewhere else, or is it an intended behavior change?

Comment (PatrykBuniX) [reply src/script/connection/ConnectionRepository.ts:168]: This is intended - blocking a conversation does not longer hide it (no need to open "next" conversation anymore).

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx (hunk 3 above)
Comment (atomrc) [reply src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: you can probably use `replaceReactComponents`. 

```
replaceReactComponents(t('otherUserNotSupportMLSMsg'), {
  start: '<bold>',
  end: '</bold>',
  render: value => <b>value</b>
}, 
```

Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]: Addressed with https://github.com/wireapp/wire-webapp/pull/17254/commits/4839d54055b8772b807073975e71dbd283b6c8dc

File: src/script/components/ParticipantItemContent/ParticipantItemContent.tsx (hunk 6 above)
Comment (PatrykBuniX) [reply src/script/components/ParticipantItemContent/ParticipantItemContent.tsx]: Thanks for catching this one @przemvs @atomrc, addressed with https://github.com/wireapp/wire-webapp/pull/17254/commits/1fc1eb1f39d4739446dc934cc2c7c45ac6938ccd

File: src/script/components/Avatar/AvatarBadge.tsx [hunk 9]
```diff
@@ -25,24 +25,36 @@ import {STATE} from '.';
...
     [STATE.PENDING]: '\\e165',
     [STATE.BLOCKED]: '\\e104',
   };
+
+  const defaultBackgroundColor = 'rgba(0, 0, 0, .56)';
+  const backgroundColor: Record<string, string> = {
+    [STATE.BLOCKED]: 'var(--white)',
+  };
+
+  const defaultColor = '#fff';
```
Comment (przemvs) [inline src/script/components/Avatar/AvatarBadge.tsx]: Try to use variables instead :)
```suggestion
  const defaultColor = 'var(--white)';
```

**Review by atomrc — APPROVED** [review]

File: src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                {
+                  exactMatch: '{{selfUserName}}',
+                  render: () => <strong>{user.name()}</strong>,
+                },
+              ])}
+            </span>
+            <>
+              {' '}
+              <Link
+                css={{fontSize: 'var(--font-size-small)', fontWeight: 600}}
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage/ReadOnlyConversationMessage.tsx]:
```suggestion
                css={{fontSize: 'var(--font-size-small)', fontWeight: 'var(--font-weight-semibold)'}}
```

**Review by przemvs — APPROVED** [review]

**Pull request merged**